Properties support for microprofile-reactive-messaging

angelozerr opened the pull request:
Properties support for microprofile-reactive-messaging

Fixes #26

Signed-off-by: azerr <azerr@redhat.com>

angelozerr commented:
I create a draft PR which supports MP reactive Messaging for properties files (with YAML we have some limitation).  This PR introduces a new mechanism which recompute available properties according value of some property.

I mean if we have:

```
mp.messaging.outgoing.generated-price.connector=smallrye-kafka
```

The following property is available:

`mp.messaging.outgoing.generated-price.topic=prices`

`topic` attribute comes from the `smallrye-kafka` connector, in other words if we have

```
mp.messaging.outgoing.generated-price.connector=smallrye-xxx
```

`mp.messaging.outgoing.generated-price.topic=prices` will appears as an error.

Here a demo:

![MPReactiveMessagingDemo](https://user-images.githubusercontent.com/1932211/89416997-628b8680-d72e-11ea-9252-527f46141921.gif)

My code is not cleaned, but you can play with this PR. This PR uses ConnectorAttribute annotation which is hosted for the moment in smallrye, so it will work only with smallrye reactive messaging implementation.

angelozerr commented:
My code is clean and I add some explanation in the README with the new extension.

datho7561 commented on microprofile.ls/README.md:
```diff
@@ -112,3 +112,26 @@ To add external snippets (like Quarkus snippets) an implementation of `ISnippetR
  * [JSON Quarkus snippet](https://github.com/redhat-developer/quarkus-ls/tree/master/quarkus.ls.ext/com.redhat.quarkus.ls/src/main/resources/com/redhat/quarkus/snippets).
  * Java Quarkus snippets loader must be declared in [META-INF/services/org.eclipse.lsp4mp.ls.commons.snippets.ISnippetRegistryLoader](https://github.com/redhat-developer/quarkus-ls/blob/master/quarkus.ls.ext/com.redhat.quarkus.ls/src/main/resources/META-INF/services/org.eclipse.lsp4mp.ls.commons.snippets.ISnippetRegistryLoader) 
+
+ Manage complex properties
```
Managing complex properties (just to be consistent with the other headers)

angelozerr replied:
fixed

datho7561 commented on microprofile.ls/README.md:
```diff
@@ -112,3 +112,26 @@ To add external snippets (like Quarkus snippets) an implementation of `ISnippetR
+ Manage complex properties
+-------
+
+The available properties in `microprofile-config.properties` comes from the external component (ex: MicroProfile JDT LS extension). In some case a property 
```
The properties available in `microprofile-config.properties` come from ...

angelozerr replied:
fixed.

datho7561 commented on microprofile.ls/README.md:
```diff
@@ -112,3 +112,26 @@ To add external snippets (like Quarkus snippets) an implementation of `ISnippetR
+The available properties in `microprofile-config.properties` comes from the external component (ex: MicroProfile JDT LS extension). In some case a property 
+cannot be computed on the external component and must be computed on MicroProfile LS side.
+
+A usecase for that, is the MicroProfile Reactive Messaging. Some properties are computed according a property value declared in the `microprofile-config.properties`. 
```
An example of this is the MicroProfile Reactive Messaging properties.

angelozerr replied:
fixed

datho7561 commented on microprofile.ls/README.md:
```diff
@@ -112,3 +112,26 @@ To add external snippets (like Quarkus snippets) an implementation of `ISnippetR
+The available properties in `microprofile-config.properties` comes from the external component (ex: MicroProfile JDT LS extension). In some case a property 
+cannot be computed on the external component and must be computed on MicroProfile LS side.
+
+A usecase for that, is the MicroProfile Reactive Messaging. Some properties are computed according a property value declared in the `microprofile-config.properties`. 
```
... according to the value of a property declared in ...

datho7561 commented on microprofile.ls/README.md:
```diff
@@ -112,3 +112,26 @@ To add external snippets (like Quarkus snippets) an implementation of `ISnippetR
+cannot be computed on the external component and must be computed on MicroProfile LS side.
+
+A usecase for that, is the MicroProfile Reactive Messaging. Some properties are computed according a property value declared in the `microprofile-config.properties`. 
+If the sample [Configuring the Kafka connector](https://quarkus.io/guides/kafka#configuring-the-kafka-connector), you have those properties
```
In the sample...

angelozerr replied:
fixed

datho7561 commented on microprofile.ls/README.md:
```diff
@@ -112,3 +112,26 @@ To add external snippets (like Quarkus snippets) an implementation of `ISnippetR
+cannot be computed on the external component and must be computed on MicroProfile LS side.
+
+A usecase for that, is the MicroProfile Reactive Messaging. Some properties are computed according a property value declared in the `microprofile-config.properties`. 
+If the sample [Configuring the Kafka connector](https://quarkus.io/guides/kafka#configuring-the-kafka-connector), you have those properties
```
... you have these properties:

angelozerr replied:
fixed

datho7561 commented on microprofile.ls/README.md:
````diff
@@ -112,3 +112,26 @@ To add external snippets (like Quarkus snippets) an implementation of `ISnippetR
+mp.messaging.outgoing.generated-price.topic=prices
+```
+
+The `topic` attribute for `mp.messaging.outgoing.generated-price.topic` comes from the `smallrye-kafka` which is configured with 
````
... comes from `smallrye-kafka` ...

datho7561 commented on microprofile.ls/README.md:
````diff
@@ -112,3 +112,26 @@ To add external snippets (like Quarkus snippets) an implementation of `ISnippetR
+```
+
+The `topic` attribute for `mp.messaging.outgoing.generated-price.topic` comes from the `smallrye-kafka` which is configured with 
+`mp.messaging.outgoing.generated-price.connector=smallrye-kafka`. In other words `mp.messaging.outgoing.generated-price.topic` exists only 
````
In other words, ...

angelozerr replied:
fixed

datho7561 commented on microprofile.ls/README.md:
```diff
@@ -112,3 +112,26 @@ To add external snippets (like Quarkus snippets) an implementation of `ISnippetR
+`mp.messaging.outgoing.generated-price.connector=smallrye-kafka`. In other words `mp.messaging.outgoing.generated-price.topic` exists only 
+if there is the declaration `mp.messaging.outgoing.generated-price.connector=smallrye-kafka` 
+
+The `mp.messaging.outgoing.generated-price.topic` property cannot be computed on external component side because it depends on the value of 
```
.. on the external component ...

angelozerr replied:
fixed

datho7561 commented on microprofile.ls/README.md:
```diff
@@ -112,3 +112,26 @@ To add external snippets (like Quarkus snippets) an implementation of `ISnippetR
+if there is the declaration `mp.messaging.outgoing.generated-price.connector=smallrye-kafka` 
+
+The `mp.messaging.outgoing.generated-price.topic` property cannot be computed on external component side because it depends on the value of 
+`mp.messaging.outgoing.generated-price.connector`. The compute is done on MicroProfile LS side with custom builder by using Java SPI [ItemMetadataProviderFactory](/src/main/java/org/eclipse/lsp4mp/extensions/ItemMetadataProviderFactory).
```
The computation ...

angelozerr replied:
fixed

datho7561 commented:
One limitation I found so far is that if the connector is configured for a stream that doesn't exist, the other properties for the invalid stream aren't marked as invalid, and are also suggested. Here, I renamed the inbound stream in the Java files for [this example Quarkus/Kafka project](https://github.com/quarkusio/quarkus-quickstarts/tree/master/kafka-quickstart):
![Screenshot from 2020-08-12 14-29-15](https://user-images.githubusercontent.com/22376627/90054136-6dec2e00-dca9-11ea-9047-6cf7b4db53bf.png)

datho7561 commented on microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/extensions/ItemMetadataProviderFactory.java:
```diff
@@ -0,0 +1,28 @@
+public interface ItemMetadataProviderFactory {
+
+	/**
+	 * Create an {@link ItemMetadata} provider factory.
```
This class is the factory, it makes an ItemMetadataProvider.

angelozerr replied:
fixed

datho7561 commented on microprofile.ls/org.eclipse.lsp4mp.ls/src/test/resources/org/eclipse/lsp4mp/extensions/reactivemessaging/mp-reactive-messaging-properties.json:
```diff
@@ -0,0 +1,80 @@
+			"sourceMethod": "generate()QFlowable\u003cQInteger;\u003e;",
+			"required": true,
+			"phase": 0,
+			"name": "mp.messaging.outgoing.generated-price.connector",
+			"sourceType": "org.acme.kafka.PriceGenerator",
```
This probably shouldn't be hardcoded; `PriceGenerator` is specific to the particular Quarkus quickstart project.

angelozerr replied:
What is the problem with that? It's just for test.

datho7561 replied:
Okay. Sorry, I couldn't tell that this was just for the tests.

datho7561 commented on microprofile.ls/org.eclipse.lsp4mp.ls/src/test/resources/org/eclipse/lsp4mp/extensions/reactivemessaging/mp-reactive-messaging-properties.json:
```diff
@@ -0,0 +1,80 @@
+			"sourceMethod": "process(I)D",
+			"required": true,
+			"phase": 0,
+			"name": "mp.messaging.outgoing.my-data-stream.connector",
+			"sourceType": "org.acme.kafka.PriceConverter",
```
This also shouldn't be hardcoded

angelozerr replied:
What is the problem with that? It's just for test.

datho7561 replied:
Okay. Sorry, I couldn't tell that this was just for the tests.

datho7561 commented on microprofile.ls/org.eclipse.lsp4mp.ls/src/test/resources/org/eclipse/lsp4mp/extensions/reactivemessaging/mp-reactive-messaging-properties.json:
```diff
@@ -0,0 +1,80 @@
+			"sourceMethod": "process(I)D",
+			"required": true,
+			"phase": 0,
+			"name": "mp.messaging.incoming.prices.connector",
+			"sourceType": "org.acme.kafka.PriceConverter",
```
This also shouldn't be hardcoded

angelozerr replied:
What is the problem with that? It's just for test.

datho7561 replied:
Okay. Sorry, I couldn't tell that this was just for the tests.

angelozerr commented on microprofile.jdt/org.eclipse.lsp4mp.jdt.test/projects/maven/microprofile-reactive-messaging/src/main/java/org/acme/GreetingResource.java:
```diff
@@ -0,0 +1,16 @@
+package org.acme;
```
fixed

angelozerr commented on microprofile.jdt/org.eclipse.lsp4mp.jdt.test/projects/maven/microprofile-reactive-messaging/src/test/java/org/acme/GreetingResourceTest.java:
```diff
@@ -0,0 +1,21 @@
+package org.acme;
```
fixed

angelozerr commented on microprofile.jdt/org.eclipse.lsp4mp.jdt.test/projects/maven/microprofile-reactive-messaging/src/test/java/org/acme/NativeGreetingResourceIT.java:
```diff
@@ -0,0 +1,9 @@
+package org.acme;
```
fixed

angelozerr commented:
> One limitation I found so far is that if the connector is configured for a stream that doesn't exist, the other properties for the invalid stream aren't marked as invalid, and are also suggested. Here, I renamed the inbound stream in the Java files for this example Quarkus/Kafka project:

Good catch! Fixed with test.

fbricon commented:
@angelozerr please make sure new file headers follow this format:

> /********************************************************************************
>  * Copyright (c) {date} {owner}[ and others]
>  * 
>  * This program and the accompanying materials are made available under the
>  * terms of the Eclipse Public License v. 2.0 which is available at
>  * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
>  * which is available at https://www.apache.org/licenses/LICENSE-2.0.
>  * 
>  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
> 
> ********************************************************************************/

angelozerr commented on microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/reactivemessaging/properties/MicroProfileReactiveMessagingProvider.java:
```diff
@@ -0,0 +1,384 @@
+		}
+	}
+
+	private static String getMPMessaggingName(MessageType messageType, boolean dynamic, String connectorOrChannelName,
```
fixed

angelozerr commented:
> @angelozerr please make sure new file headers follow this format:

fixed

angelozerr commented:
> Since the name of the consumed channel must not be blank in order to conform to the API, neither of these should be suggested.

Good cacth, fixed.
